### PR TITLE
(misc) Create a agent testing harness

### PR DESCRIPTION
**Pull request by ripienaar**
This adds a harness that can be used to host
agents in a running Server instance that has
their actions stubable using gomock

Signed-off-by: R.I.Pienaar <rip@devco.net>

### CODE REVIEW

File: integration/suites/harness/agent_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+		_, rpcutilClient, err = createRpcUtilClient()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		mockCtl.Finish()
+	})
+
+	Describe("action stubbing", func() {
```
Comment (ripienaar) [inline integration/suites/harness/agent_test.go:131]: @vjanelle the beef of things here, after in the BeforeEach() adding the test harness into the server pretending to be rpcutil we call it here using the unmodified generated rpcutil client and doing mocks on that `ping` action.

File: integration/suites/harness/agent_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		It("Should do the correct stubs", func() {
+			rpcutilAgent.Stub("ping", func(_ context.Context, _ *mcorpc.Request, reply *mcorpc.Reply, _ *mcorpc.Agent, _ inter.ConnectorInfo) {
+				reply.Data = map[string]interface{}{
+					"pong": time.Now().Unix(),
+				}
+			}).AnyTimes()
+
+			for i := 0; i < 5; i++ {
+				res, err := rpcutilClient.OptionTargets([]string{"localhost"}).Ping().Do(ctx)
```
Comment (vjanelle) [inline integration/suites/harness/agent_test.go:140]: Only thing I couldn't figure out, I tried OptionsFactFilter (after populating testdata/facts.yaml)  and it wouldn't discover any results.

Comment (ripienaar) [reply integration/suites/harness/agent_test.go:140]:
```
		da, err := discovery.New(srv.AgentManager())
		Expect(err).ToNot(HaveOccurred())
		err=srv.AgentManager().RegisterAgent(ctx, "discover", da, srv.Connector())
		Expect(err).ToNot(HaveOccurred())
```

should sort it, add to `startServerInstance()`

Comment (ripienaar) [reply integration/suites/harness/agent_test.go:140]: The way discovery works is a filtered message is sent to the `discovery` agent with just "ping", and if the server filter validation worked a "pong" is sent back.

https://github.com/choria-io/go-choria/blob/dcc5c3fdb8bbaf2eb736c50c28325e47755afd59/providers/agent/mcorpc/golang/discovery/discovery.go#L57-L72

Agents dont need to be "RPC" agents as such, any payload is supported in any format and any serialization. RPC agents formalize the payload to make a RPC request, but discovery doesnt need all that so just string ping.

**Review by vjanelle — APPROVED** [review]

**Pull request merged**